crates/sui-core/src/streamer.rs

```diff
@@ -1,21 +1,96 @@
 // Copyright (c) 2022, Mysten Labs, Inc.
 // SPDX-License-Identifier: Apache-2.0
-use sui_types::event::EventEnvelope;
-use tokio::{sync::mpsc::Receiver, task::JoinHandle};
-use tracing::debug;
 
-pub struct Streamer {
-    event_queue: Receiver<EventEnvelope>,
+use crate::event_handler::EVENT_DISPATCH_BUFFER_SIZE;
+use futures::Stream;
+use std::collections::BTreeMap;
+use std::fmt::Debug;
+use std::sync::Arc;
+use sui_types::base_types::ObjectID;
+use sui_types::error::SuiError;
+use tokio::runtime::Handle;
+use tokio::sync::mpsc::Sender;
+use tokio::sync::{mpsc, RwLock};
+use tokio_stream::wrappers::ReceiverStream;
+use tracing::{debug, warn};
+
+type Subscribers<T> = BTreeMap<String, Sender<T>>;
+type FilterMap<F, T> = Arc<RwLock<BTreeMap<F, T>>>;
+pub struct Streamer<T, F: Filter<T>> {
+    streamer_queue: Sender<T>,
+    subscribers: FilterMap<F, Subscribers<T>>,
 }
 
-impl Streamer {
-    pub fn spawn(rx: Receiver<EventEnvelope>) -> JoinHandle<()> {
-        tokio::spawn(async move { Self { event_queue: rx }.stream().await })
+impl<T, F> Streamer<T, F>
+where
+    T: Clone + Debug + Send + Sync + 'static,
+    F: Filter<T> + Ord + Clone + Send + Sync + 'static + Clone,
+{
+    pub fn spawn(buffer: usize) -> Self {
+        let (tx, rx) = mpsc::channel::<T>(buffer);
+        let streamer = Self {
+            streamer_queue: tx,
+            subscribers: Default::default(),
+        };
+        let mut rx = rx;
+
+        let subscribers = streamer.subscribers.clone();
+        tokio::spawn(async move {
+            while let Some(data) = rx.recv().await {
+                Self::send_to_all_subscribers(subscribers.clone(), data).await;
+            }
+        });
+        streamer
     }
 
-    pub async fn stream(&mut self) {
-        while let Some(event_envelope) = self.event_queue.recv().await {
-            debug!(event_envelope =? event_envelope, "Get event");
+    async fn send_to_all_subscribers(filter_map: FilterMap<F, Subscribers<T>>, data: T) {
+        for (filter, subscribers) in filter_map.clone().read().await.clone() {
+            if !(filter.matches(&data)) {
+                continue;
+            }
+            for (id, subscriber) in subscribers {
+                let data = data.clone();
+                let filter_map = filter_map.clone();
```

Comment:
Why is this needed?   This is quite expensive, to clone the filter map for every message

Reply:
this is needed to move it into the async block, filter_map is Arc<Mutex<_>>, we are just cloning the pointer here